Adding pre and post update tick methods to core engine.

ambientenergy102 opened the pull request:
Added pre and post ticks to the engine to support additional update group functionality.
Added manual update flags for allowing nodes to tick their own update methods in gdscript after their C++ tick to support base classes running logic before gdscripts get a chance to touch it.

Please take a look at the terminology and critique it!

== Testing
Tested with halcyon-one project in debug / release_debug in editor and standalone.

ambientenergy102 committed:
Adding pre and post update tick methods to core engine.

npeaks reviewed (approved):
this seems pretty straight forward, I'm a fan. I'd like Francois to look as well though, there might be a consideration I'm missing (not as familiar with the inner workings of the godot engine). 

If anything, I might call it `pre_process` instead of `pre_update`, just to be consistent with the existing godot terminology.

ambientenergy102 commented:
> If anything, I might call it `pre_process` instead of `pre_update`, just to be consistent with the existing godot terminology.

Agree

Razoric480 commented on scene/main/node.h:
```diff
@@ -280,6 +289,14 @@ class Node : public Object {
+		NOTIFICATION_PRE_UPDATE_INTERNAL_PROCESS = 29,
+		NOTIFICATION_PRE_UPDATE_PROCESS = 30,
+		NOTIFICATION_POST_UPDATE_INTERNAL_PROCESS = 31,
+		NOTIFICATION_POST_UPDATE_PROCESS = 32,
```
These should also be added as constants in `Node::_bind_methods()`

Razoric480 commented on scene/main/node.h:
```diff
@@ -393,6 +410,14 @@ class Node : public Object {
+	void Node::set_post_process(bool p_pre_process);
+	void Node::set_manual_process(bool manual);
+	void Node::set_manual_pre_process(bool manual);
+	void Node::set_manual_post_process(bool manual);
```
Likewise, these should be added to _bind_methods() so they can be accessed via scripting.

Razoric480 replied:
Also maybe some commensurate `is_`, so it's like set_process/is_processing

Razoric480 commented on core/os/main_loop.cpp:
```diff
@@ -45,6 +45,10 @@ void MainLoop::_bind_methods() {
 	BIND_VMETHOD(MethodInfo(Variant::BOOL, "_idle", PropertyInfo(Variant::REAL, "delta")));
+	//## BEGIN_ENGINE_EDIT
+	BIND_VMETHOD(MethodInfo("_pre_update", PropertyInfo(Variant::REAL, "delta")));
+	BIND_VMETHOD(MethodInfo("_post_update", PropertyInfo(Variant::REAL, "delta")));
```
I'd also like to see this added to Node's _bind_methods(), alongside _ready, _process, _physics_process, etc as VMETHODs

ambientenergy102 replied:
@Razoric480 
@npeaks 
I've updated this PR with fixes as outlined.

Razoric480 reviewed (changes requested):
Looks good - my main complaint is the lack of access for Node to have a pre/post update function that's exposed as a virtual method in scripts; instead it's just implied via `call_multilevel`, so we should open the doors for scripts to use it.

Also can you take the following steps?
```bash
cd godot/src/godot
bin/<godot binary> --doctool .
```
this will generate new entries in the XML for Engine, MainLoop, Node and SceneTree (and some of our custom modules but you can discard those changes) that I think we could stand to update with these new changes' documentation.

ambientenergy102 committed:
Updating core engine to standardize pre-process and post-process names.
Adding vmethods for pre/post process
Adding helper methods for script access to node for new functionality
Updating gitignore to remove sentry & other visual studio intermediate files
Updated node and main loop documentation.

Razoric480 reviewed (approved):
Thanks for that

Though if you get a chance, can you edit the <description> field of those XML files. May as well add some documentation about what they're for in there.

Though we have a whole mess of XML files we have to fill out from all our modules, so we should do a documentation pass over it at some point.

ambientenergy102 commented:
>
I'll have to hit you up later about the doctool, it generates a lot of files... might want to do a separate PR on it to make sure its proper as there are about 20 new ones (sans the obvious game module updates)

ambientenergy102 committed:
Adding comments to MainLoop & Node documentation.
